### ISSUE TITLE
Error with netty

### ISSUE THREAD

**Issue by OcelusPRO** [crackthecodeabhi/kreds#80]
Using kreds occasionally generates the following error:
```
[2023-03-03 04:11:12] [nioEventLoopGroup-2-1] [io.netty.util.ResourceLeakDetector] ERROR - LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.AbstractByteBufAllocator.compositeDirectBuffer(AbstractByteBufAllocator.java:224)
	io.netty.buffer.AbstractByteBufAllocator.compositeBuffer(AbstractByteBufAllocator.java:202)
	io.netty.handler.codec.MessageAggregator.decode(MessageAggregator.java:269)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:831)

```

**Comment by crackthecodeabhi** [comment]
thank @OcelusPRO will look into it.
Can you provide more info on how you are using it.

Any hints would be helpful to reproduce the isssue, like 
what commands you are using in your application?
Are you using pub/sub?

Thanks

**Comment by OcelusPRO** [comment]
Here is the part of my code in which I use the library:

```kt
package fr.ftnl.bot.modules.cache

import fr.ftnl.bot.CONFIG
import fr.ftnl.bot.GSON
import fr.ftnl.bot.modules.cache.RedisManager.client
import io.github.crackthecodeabhi.kreds.args.SyncOption
import io.github.crackthecodeabhi.kreds.connection.Endpoint
import io.github.crackthecodeabhi.kreds.connection.KredsClient
import io.github.crackthecodeabhi.kreds.connection.newClient
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

/**
 * Redis manager.
 * @property client [KredsClient] KRedis client.
 */
object RedisManager {
    private val host = CONFIG.redisConfiguration.host
    val client: KredsClient = newClient(Endpoint.from(host))
    init {
        runBlocking {
            launch {
                val password = CONFIG.redisConfiguration.password
                if (password.isNotEmpty()) {
                    client.auth(password)
                }
                val db = CONFIG.redisConfiguration.database
                client.select(db)
            }
        }
    }
    
    /**
     * Getters.
     * @see [RedisManager]
     */
    object Getters {
        
        /**
         * Get a value from the Redis database.
         * @param key [String] Key of the value.
         *
         * @return [T] Value.
         */
        suspend inline fun <reified T> get(key: String): T? = GSON.fromJson(client.get(key), T::class.java)
        
        /**
         * Get all values from the Redis database.
         *
         * @param pattern [String] Pattern to match.
         *
         * @return [Map] Map of keys and values.
         */
        suspend fun getAll(pattern: String = "*") = client.keys(pattern).associateWith { client.get(it) }
        
    }
    
    /**
     * Setters.
     */
    object Setters {
        /**
         * Set a value in the Redis database.
         *
         * @param key [String] Key of the value.
         * @param objects [T] Value.
         *
         * @return [T] Value.
         */
        suspend fun <T> set(key: String, objects: T): T{
            client.set(key, GSON.toJson(objects))
            return objects
        }
        
        /**
         * Set a value in the Redis database with an expiration time.
         *
         * @param key [String] Key of the value.
         * @param objects [T] Value.
         * @param expireSeconds [Long] Expiration time in seconds.
         *
         * @return [T] Value.
         */
        suspend fun <T> set(key: String, objects: T, expireSeconds: Long): T{
            client.set(key, GSON.toJson(objects))
            client.expire(key, expireSeconds.toULong())
            return objects
        }
    }
    
    /**
     * Deletion.
     */
    object Deletion {
        /**
         * Delete a value from the Redis database.
         *
         * @param key [String] Key of the value.
         *
         * @return [Long] Number of deleted keys.
         */
        suspend fun del(key: String) = client.del(key)
        
        /**
         * Delete all values from the Redis database matching pattern.
         *
         * @param pattern [String] Pattern to match.
         */
        suspend fun invalidateAll(pattern: String) = client.keys(pattern).forEach { client.del(it) }
        
        /**
         * Delete all values from the Redis database.
         */
        suspend fun invalidateAll() = client.flushAll(SyncOption.SYNC)
    }
    
    /**
     * Int manipulation.
     */
    object IntManipulation {
        /**
         * Increment a value in the Redis database.
         *
         * @param key [String] Key of the value.
         *
         * @return [Long] new value.
         */
        suspend fun incr(key: String): Long = client.incr(key)
        
        /**
         * Decrement a value in the Redis database.
         *
         * @param key [String] Key of the value.
         *
         * @return [Long] new value.
         */
        suspend fun decr(key: String): Long = client.decr(key)
    }
    
    /**
     * Time remaining.
     */
    object TimeRemaining {
        /**
         * Get the time remaining before a key expires.
         *
         * @param key [String] Key of the value.
         *
         * @return [Long] Time remaining in seconds.
         */
        suspend fun ttl(key: String): Long = client.ttl(key)
        
        /**
         * Get the time remaining before a key expires.
         *
         * @param key [String] Key of the value.
         *
         * @return [Long] Time remaining in milliseconds.
         */
        suspend fun pttl(key: String): Long = client.pttl(key)
    }
}
```

The following methods are used by my program:
rtl, incr, invalidateAll(pattern), the different sets and get/getAll(pattern)

I don't currently use the pub/sub

It seems to me that the error occurs either when I do too many sets in a row, or in the invalidateAll(pattern)

Sorry for not being able to be more precise, the error itself is not practical to understand

**Comment by lpawlak1** [comment]
Issue also exists when using only `hset`/`hgetall`/`hdel`

**Comment by ageorgousakis** [comment]
I get the same errors occusionally using Kreds with Ktor. Running using the following system property `-Dio.netty.leakDetectionLevel=advanced` I get more detailed exception  like that:

```
2023-11-04 14:52:37.034 [pGroupProxy-5-1] TRACE i.g.c.kreds.connection.KonnectionImpl    - New connection created to Endpoint(host=localhost, port=6379)
...

2023-11-04 14:52:37.824 [pGroupProxy-4-2] ERROR io.netty.util.ResourceLeakDetector       - LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
#1:
	io.netty.buffer.AdvancedLeakAwareCompositeByteBuf.toString(AdvancedLeakAwareCompositeByteBuf.java:728)
	io.github.crackthecodeabhi.kreds.UtilsKt.toDefaultCharset(Utils.kt:43)
	io.github.crackthecodeabhi.kreds.protocol.BulkStringHandler.doHandle(MessageHandlers.kt:56)
	io.github.crackthecodeabhi.kreds.protocol.BulkStringHandler.doHandle(MessageHandlers.kt:49)
	io.github.crackthecodeabhi.kreds.protocol.CommandProcessor.decode(Command.kt:72)
	io.github.crackthecodeabhi.kreds.connection.AbstractKredsClient$execute$suspendImpl$$inlined$withReentrantLock$2.invokeSuspend(ExclusiveObject.kt:49)
	kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.ktor.server.netty.EventLoopGroupProxy$Companion.create$lambda$1$lambda$0(NettyApplicationEngine.kt:296)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:842)
#2:
	Hint: 'KonnectionImpl$ResponseHandler#0' will handle the message from this point.
	io.netty.buffer.AdvancedLeakAwareCompositeByteBuf.touch(AdvancedLeakAwareCompositeByteBuf.java:36)
	io.netty.buffer.DefaultByteBufHolder.touch(DefaultByteBufHolder.java:105)
	io.netty.handler.codec.redis.FullBulkStringRedisMessage.touch(FullBulkStringRedisMessage.java:228)
	io.netty.handler.codec.redis.FullBulkStringRedisMessage.touch(FullBulkStringRedisMessage.java:27)
	io.netty.channel.DefaultChannelPipeline.touch(DefaultChannelPipeline.java:116)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:417)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:842)
#3:
	io.netty.buffer.AdvancedLeakAwareCompositeByteBuf.retain(AdvancedLeakAwareCompositeByteBuf.java:36)
	io.netty.buffer.DefaultByteBufHolder.retain(DefaultByteBufHolder.java:87)
	io.netty.handler.codec.redis.FullBulkStringRedisMessage.retain(FullBulkStringRedisMessage.java:210)
	io.netty.handler.codec.redis.FullBulkStringRedisMessage.retain(FullBulkStringRedisMessage.java:27)
	io.netty.util.ReferenceCountUtil.retain(ReferenceCountUtil.java:41)
	io.netty.handler.codec.redis.RedisArrayAggregator.decode(RedisArrayAggregator.java:46)
	io.netty.handler.codec.redis.RedisArrayAggregator.decode(RedisArrayAggregator.java:33)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:842)
#4:
	Hint: 'RedisArrayAggregator#0' will handle the message from this point.
	io.netty.buffer.AdvancedLeakAwareCompositeByteBuf.touch(AdvancedLeakAwareCompositeByteBuf.java:36)
	io.netty.buffer.DefaultByteBufHolder.touch(DefaultByteBufHolder.java:105)
	io.netty.handler.codec.redis.FullBulkStringRedisMessage.touch(FullBulkStringRedisMessage.java:228)
	io.netty.handler.codec.redis.FullBulkStringRedisMessage.touch(FullBulkStringRedisMessage.java:27)
	io.netty.channel.DefaultChannelPipeline.touch(DefaultChannelPipeline.java:116)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:417)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:842)
#5:
	io.netty.buffer.AdvancedLeakAwareCompositeByteBuf.addComponent(AdvancedLeakAwareCompositeByteBuf.java:920)
	io.netty.handler.codec.MessageAggregator.appendPartialContent(MessageAggregator.java:331)
	io.netty.handler.codec.MessageAggregator.decode(MessageAggregator.java:296)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:842)
Created at:
	io.netty.buffer.AbstractByteBufAllocator.compositeDirectBuffer(AbstractByteBufAllocator.java:224)
	io.netty.buffer.AbstractByteBufAllocator.compositeBuffer(AbstractByteBufAllocator.java:202)
	io.netty.handler.codec.MessageAggregator.decode(MessageAggregator.java:269)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:842)
: 1 leak records were discarded because they were duplicates
: 2 leak records were discarded because the leak record count is targeted to 4. Use system property io.netty.leakDetection.targetRecords to increase the limit.
```

**Comment by crackthecodeabhi** [comment]
@ageorgousakis were you able to consistently reproduce this error?

**Comment by ageorgousakis** [comment]
I created a small application to check it

https://github.com/ageorgousakis/ktor-kreds-sample

Run the application using option `-Dio.netty.leakDetectionLevel=advanced` to see more detail log and use a tool to issue parallel request to see the problem

```shell
bombardier -c 3 -d 5s -l -H "Content-Type: application/json" -H "Connection: keep-alive" http://localhost:8080/users/1
```

**Comment by crackthecodeabhi** [comment]
@ageorgousakis thanks, will take a look, that was helpful.

**Comment by pihanya** [comment]
I have the same problem. Please fix it

**Comment by crackthecodeabhi** [comment]
@pihanya @ageorgousakis @lpawlak1 @OcelusPRO  Good news, i have found the fix! I will run some tests and release it!

**Comment by crackthecodeabhi** [comment]
Fixed, tested and realease in [v0.9.1](https://github.com/crackthecodeabhi/kreds/releases/tag/0.9.1)
